File: ts/game-of-life/src/grid.ts
import { DeadCell, LiveCell } from "./__tests__/gameOfLife.test";

export class Grid {
    private grid: (DeadCell | LiveCell)[][];

    constructor(rows: number, cols: number) {
        this.grid = Array.from({ length: rows }, () => Array(cols).fill(new DeadCell()));
    }

    public countNeighbours(row: number, col: number): number {
        let count = 0;
        for (let i = row - 1; i <= row + 1; i++) {
            for (let j = col - 1; j <= col + 1; j++) {
                if (i >= 0 && i < this.grid.length && j >= 0 && j < this.grid[0].length) {
                    if (this.grid[i][j] instanceof LiveCell) {
                        count += 1;
                    }
                }
            }
        }
        return count;
    }

    public setCell(row: number, col: number, value: DeadCell | LiveCell): void {
        this.grid[row][col] = value;
    }

    public getCell(row: number, col: number): DeadCell | LiveCell {
        return this.grid[row][col];
    }
    public print(): void {
        console.log(this.grid.map(row => row.join(' ')).join('\n'));
    }
}

File: csharp/GameOfLife/GameOfLifeApp/Grid.cs
using System;

namespace GameOfLifeApp
{
    public class Grid
    {
        public Cell[,] cells;

        public Grid(int rows, int cols)
        {
            cells = new Cell[rows, cols];

            for (int i = 0; i < cells.GetLength(0); i++)
            {
                for (int j = 0; j < cells.GetLength(1); j++)
                {
                    cells[i, j] = new DeadCell();
                }
            }
        }

        public void NextGeneration()
        {
            Cell[,] newGrid = new Cell[cells.GetLength(0), cells.GetLength(1)];

            for (int i = 0; i < cells.GetLength(0); i++)
            {
                for (int j = 0; j < cells.GetLength(1); j++)
                {
                    int nbNeighbours = CountNeighbours(i, j);
                    newGrid[i, j] = cells[i, j].NextGeneration(nbNeighbours);
                }
            }

            cells = newGrid;
        }

        public int CountNeighbours(int i, int j)
        {
            int count = 0;

            for (int x = i - 1; x <= i + 1; x++)
            {
                for (int y = j - 1; y <= j + 1; y++)
                {
                    if (
                        x >= 0
                        && x < cells.GetLength(0)
                        && y >= 0
                        && y < cells.GetLength(1)
                        && !(x == i && y == j)
                    )
                    {
                        if (cells[x, y] is LiveCell)
                        {
                            count++;
                        }
                    }
                }
            }

            return count;
        }

        public void Print()
        {
            for (int i = 0; i < cells.GetLength(0); i++)
            {
                for (int j = 0; j < cells.GetLength(1); j++)
                {
                    Console.Write(cells[i, j] + " ");
                }
                Console.WriteLine();
            }
        }
    }
}
